packages/binding-wotfirestore/src/wotfirestore-handler.ts

```diff
@@ -0,0 +1,296 @@
+      !fbConfig ||
+      !fbConfig.user ||
+      !fbConfig.user.email ||
+      !fbConfig.user.password
+    ) {
+      throw new Error('firebase auth error: cannot find email/password')
+    }
+    const firestore = await new Promise((resolve, reject) => {
+      firebase
+        .auth()
+        .signInWithEmailAndPassword(fbConfig.user.email, fbConfig.user.password)
+        .then(() => {
+          const firestore = firebase.firestore()
+          resolve(firestore)
+        })
+        .catch(function (error) {
+          reject(`firebase auth error: ${error}`)
+        })
+    })
+    return firestore
+  } else {
+    return firebase.firestore()
+  }
+}
+
+export const writeDataToFirestore = (
+  firestore,
+  topic: string,
+  content: Content,
+  reqId = null
+): Promise<any> => {
+  return new Promise((resolve, reject) => {
+    console.debug(
+      '[debug] writeDataToFirestore topic:',
+      topic,
+      ' value:',
+      content,
+      reqId
+    )
+    const ref = firestore.collection('things').doc(encodeURIComponent(topic))
+    let data = { updatedTime: Date.now(), reqId }
+    if (content) {
+      data['content'] = JSON.stringify(content)
+    }
+    ref
+      .set(data)
+      .then((value) => {
+        resolve(value)
+      })
+      .catch((err) => {
+        console.error(
+          '[error] failed to write data to firestore: ',
+          err,
+          ' topic: ',
+          topic,
+          ' data: ',
+          data
+        )
+        reject(err)
+      })
+  })
+}
+
+export const readDataFromFirestore = (
+  firestore,
+  topic: string
+): Promise<Content> => {
+  return new Promise<Content>((resolve, reject) => {
+    console.debug('[debug] readDataFromFirestore topic:', topic)
+    const ref = firestore.collection('things').doc(encodeURIComponent(topic))
+    ref
+      .get()
+      .then((doc) => {
+        if (doc.exists) {
+          const data = doc.data()
+          let content: Content = null
+          console.debug('[debug] readDataToFirestore gotten data:', data)
+          if (data && data.content) {
+            let obj: any = JSON.parse(data.content)
+            if (!obj) {
+              reject(new Error(`invalid ${topic} content:${content}`))
+            }
+            content = {
+              type: obj.type,
+              body:
+                obj && obj.body && obj.body.type === 'Buffer'
+                  ? Buffer.from(obj.body.data)
+                  : Buffer.from('')
+            }
+          }
+          resolve(content)
+        } else {
+          reject('no contents')
+          console.debug(
+            '[debug] read data from firestore but no contents topic:',
+            topic
+          )
+        }
+      })
+      .catch((err) => {
+        console.error(
+          '[error] failed read data from firestore: ',
+          err,
+          ' topic: ',
+          topic
+        )
+        reject(err)
+      })
+  })
+}
+
+export const subscribeFromFirestore = async (
```

Comment:
Should it be called `subscribeToFirestore` ?

Reply:
Thank you. I will modify this method name.